Source/WebCore/html/CanvasNoiseInjection.cpp

```diff
@@ -0,0 +1,229 @@
+            && bytes[index1 + 3] <= bytes[index2 + 3] && bytes[index2 + 3] <= bytes[index3 + 3];
+    };
+
+    const auto compareColorsAndSetBounds = [&areColorsDescending](const auto& bytes, auto& tightestBoundingIndices, auto& tightestBoundingDiff, auto colorIndex, auto neighborIndex1, auto neighborIndex2) {
+        if (areColorsDescending(neighborIndex1, colorIndex, neighborIndex2)) {
+            tightestBoundingIndices = { neighborIndex1, neighborIndex2 };
+            setTightnessBounds(bytes, tightestBoundingDiff, neighborIndex1, neighborIndex2);
+        } else if (areColorsDescending(neighborIndex2, colorIndex, neighborIndex1)) {
+            tightestBoundingIndices = { neighborIndex2, neighborIndex1 };
+            setTightnessBounds(bytes, tightestBoundingDiff, neighborIndex2, neighborIndex1);
+        }
+    };
+
+    if (isInTopRow || isInBottomRow) {
+        if (areColorsDescending(leftIndex, index, rightIndex))
+            return { { leftIndex, rightIndex } };
+        if (areColorsDescending(rightIndex, index, leftIndex))
+            return { { rightIndex, leftIndex } };
+        return std::nullopt;
+    }
+
+    if (isInLeftColumn || isInRightColumn) {
+        if (areColorsDescending(aboveIndex, index, belowIndex))
+            return { { aboveIndex, belowIndex } };
+        if (areColorsDescending(belowIndex, index, aboveIndex))
+            return { { belowIndex, aboveIndex } };
+        return std::nullopt;
+    }
+
+    std::optional<std::pair<int, int>> tightestBoundingIndices;
+    std::array<int, 4> tightestBoundingDiff { 255, 255, 255, 255 };
+
+    compareColorsAndSetBounds(bytes, tightestBoundingIndices, tightestBoundingDiff, index, leftIndex, rightIndex);
+    compareColorsAndSetBounds(bytes, tightestBoundingIndices, tightestBoundingDiff, index, aboveIndex, belowIndex);
+    compareColorsAndSetBounds(bytes, tightestBoundingIndices, tightestBoundingDiff, index, aboveLeftIndex, belowRightIndex);
+    compareColorsAndSetBounds(bytes, tightestBoundingIndices, tightestBoundingDiff, index, aboveRightIndex, belowLeftIndex);
+
+    return tightestBoundingIndices;
+}
+
+void CanvasNoiseInjection::postProcessDirtyCanvasBuffer(ImageBuffer* imageBuffer, NoiseInjectionHashSalt salt)
+{
+    if (m_postProcessDirtyRect.isEmpty())
+        return;
+
+    if (!imageBuffer)
+        return;
+
+    PixelBufferFormat format { AlphaPremultiplication::Unpremultiplied, PixelFormat::RGBA8, imageBuffer->colorSpace() };
+    auto pixelBuffer = imageBuffer->getPixelBuffer(format, m_postProcessDirtyRect);
+    if (!is<ByteArrayPixelBuffer>(pixelBuffer))
+        return;
+
+    if (postProcessPixelBufferResults(*pixelBuffer, salt)) {
+        imageBuffer->putPixelBuffer(*pixelBuffer, m_postProcessDirtyRect, m_postProcessDirtyRect.location());
+        m_postProcessDirtyRect = { };
+    }
+}
+
+bool CanvasNoiseInjection::postProcessPixelBufferResults(Ref<PixelBuffer>&& pixelBuffer, NoiseInjectionHashSalt salt) const
```

Comment:
Here we're saying "this method may take ownership of the reference to the pixel buffer."

The method actually never takes the ownership of the reference.
So the signature should just be` PixelBuffer&` which means "this method uses the pixel buffer".